rns_shell/compositor/layers/Layer.cpp

```diff
@@ -52,29 +64,94 @@ void Layer::appendChild(SharedLayer child) {
+        removeChild(child, index);
+        return;
+    }
+}
+
+void Layer::removeFromParent() {
+  if (parent_)
+    parent_->removeChild(this);
+}
+
+void Layer::preRoll(PaintContext& context, bool forceLayout) {
+    // Layer Layout has changed or parent has forced Layout change for child. Need to recalculate absolute frame and update absFrame_
+    if(forceLayout || (invalidateMask_ & LayerLayoutInvalidate)) {
+        // Adjust absolute position
+        //TODO Once we have skia transform matrix, need to first update frame_ i.e before calculating newAbsFrame
+        // layer_->setFrame( trnsfromed x, y width hight
+        int frameAbsX = frame_.x(), frameAbsY = frame_.y();
+        if(parent_) {
+            frameAbsX += parent_->absFrame_.x();
+            frameAbsY += parent_->absFrame_.y();
+        }
+
+        SkIRect newAbsFrame = SkIRect::MakeXYWH(frameAbsX, frameAbsY, frame_.width(), frame_.height());
+#if USE(RNS_SHELL_PARTIAL_UPDATES)
+        if((invalidateMask_ & LayerLayoutInvalidate)) {
+            // Add previous frame to damage rect only if layer layout was invalidated and new layout is different from old layout
+            if(context.supportPartialUpdate && (invalidateMask_ & LayerLayoutInvalidate)) {
+                if(absFrame_.isEmpty() != true && newAbsFrame.contains(absFrame_) != true ) {
+                    addDamgeRect(context, absFrame_); // Previous frame rect
+                    RNS_LOG_DEBUG("New frame layout is different from previous frame. Add to damage rect..");
+                }
+            }
+            RNS_LOG_DEBUG("PreRoll Layer(ID:" << layerId_ << ", ParentID:" << (parent_ ? parent_->layerId() : -1) <<
+                ") Frame [" << frame_.x() << "," << frame_.y() << "," << frame_.width() << "," << frame_.height() <<
+                "], AbsFrame(Prev,New) ([" << absFrame_.x() << "," << absFrame_.y() << "," << absFrame_.width() << "," << absFrame_.height() << "]" <<
+                " - [" << newAbsFrame.x() << "," << newAbsFrame.y() << "," << newAbsFrame.width() << "," << newAbsFrame.height() << "])");
+        }
+#endif
+        absFrame_ = newAbsFrame; // Save new absFrame
+    }
+
+#if USE(RNS_SHELL_PARTIAL_UPDATES)
+    if(invalidateMask_ != LayerInvalidateNone) {// As long as there is some invalidation , it creates a dirty rect.
+        RNS_LOG_DEBUG("AddDamage Layer(ID:" << layerId_ <<
+            ") AbsFrame[" << absFrame_.x() << "," << absFrame_.y() << "," << absFrame_.width() << "," << absFrame_.height() << "]");
+        addDamgeRect(context, absFrame_); // Previous frame rect
+    }
+#endif
+}
+
+void Layer::prePaint(PaintContext& context, bool forceLayout) {
+    //Adjust absolute Layout frame and dirty rects
+    preRoll(context, forceLayout);
+    invalidateMask_ = LayerInvalidateNone;
+
+    // prePaint children recursively
+    for (auto& layer : children()) {
+        layer->prePaint(context, (forceLayout || (invalidateMask_ & LayerLayoutInvalidate)));
```

Comment:
With invalidateMask_ = LayerInvalidateNone;, will this case satisfy "(invalidateMask_ & LayerLayoutInvalidate)"

Reply:
Good catch, Will change